Added Manual Priority

MANDAL0R3 opened the pull request:
* Removed right click to cancel construction job
* Added Manual Priority for cancel construction job
* Added Manual Priority for Tile, and Furniture
* Added Property in Job called IsBeingWorked
* Fixed a typo in ContextMenuAction
* Changed two public fields in ContextMenuAction to properites

NOTE: If a character is told to prioritize a different job and they are carrying materials, if the new job does not require the materials then they will the drop the materials on the characters current tile.(This of course, should be changed later when DumpExcessInventory is complete)


![image](https://cloud.githubusercontent.com/assets/12440052/18001617/c094cad6-6b51-11e6-8599-91d0e3058411.png)


![image](https://cloud.githubusercontent.com/assets/12440052/18001520/5f3de51a-6b51-11e6-9b38-f181bea47449.png)

Randshot commented:
This works really well. Loving this change :)
Haven't looked into the code yet.

Randshot commented on Assets/Scripts/Models/Character.cs:
```diff
@@ -874,8 +876,60 @@ public IEnumerable<ContextMenuAction> GetContextMenuActions(ContextMenu contextM
+        j.IsBeingWorked = true;
+        //check if the character is carrying any materials and if they could be used for the new job,
+        //if the character is carrying materials but is not used in the new job, then drop them
+        //on the current tile for now.
```
Between comment slashes '//' and comment should be a white space.
Also, comments should start with a capital letter.

Randshot commented on Assets/Scripts/Models/Character.cs:
```diff
@@ -874,8 +876,60 @@ public IEnumerable<ContextMenuAction> GetContextMenuActions(ContextMenu contextM
+
+        myJob = j;
+
+        // Get our destination from the job
```
End with a period.

Randshot commented on Assets/Scripts/Models/Character.cs:
```diff
@@ -874,8 +876,60 @@ public IEnumerable<ContextMenuAction> GetContextMenuActions(ContextMenu contextM
+        // Get our destination from the job
+        DestTile = myJob.tile;
+
+        // If the dest tile does not have neighbours that are walkable it's very likable that they can't be walked to
```
The same for this one.

Randshot commented on Assets/Scripts/Models/Tile.cs:
```diff
@@ -22,7 +22,7 @@ public enum ENTERABILITY
 
 [MoonSharpUserData]
-public class Tile :IXmlSerializable, ISelectable
+public class Tile :IXmlSerializable, ISelectable,IContextActionProvider
```
Maybe add a space between comma and class name?

Randshot commented on Assets/Scripts/Models/Tile.cs:
```diff
@@ -366,4 +366,27 @@ public string GetJobDescription()
+                    Action = (cm, c) => { c.PrioritizeJob(PendingBuildJob); }
+                };
+            }
+
```
I know this is nitpicking but maybe remove this empty line?

Grenkin1988 commented on Assets/Scripts/Models/Character.cs:
```diff
@@ -874,8 +876,60 @@ public IEnumerable<ContextMenuAction> GetContextMenuActions(ContextMenu contextM
         };
     }
+
+    public void PrioritizeJob(Job j)
```
Please change variable "j" name to "job"

Grenkin1988 commented on Assets/Scripts/Models/Character.cs:
```diff
@@ -874,8 +876,60 @@ public IEnumerable<ContextMenuAction> GetContextMenuActions(ContextMenu contextM
+        DestTile = myJob.tile;
+
+        // If the dest tile does not have neighbours that are walkable it's very likable that they can't be walked to
+        if (DestTile.GetNeighbours().Any((tile) => { return tile.MovementCost > 0; }) == false)
```
I gues this could be writen as:
`if (DestTile.GetNeighbours().Any(tile => tile.MovementCost > 0) == false)`
or
`if (!DestTile.GetNeighbours().Any(tile => tile.MovementCost > 0))`

Grenkin1988 commented on Assets/Scripts/Models/ContextMenu/ContextMenuAction.cs:
```diff
@@ -2,7 +2,7 @@
 public class ContextMenuAction
 {
-    public bool RequiereCharacterSelected;
+    public bool RequireCharacterSelected;
```
No public fields in class

MANDAL0R3 replied:
The public field was there long before but i guess i can change it to a property. Sorry for all of the stylecop errors but im pretty sure if i ran it on the files then i would be in for another few hours just fixing the whole file.

MANDAL0R3 commented:
Sorry for a few of the stylecop errors. I was mostly just getting it done near the end because i ran into a very very weird issue with my repository and visual studio. I was in one the files and i press undo, then next thing i knew half of the files that i changed reverted back to before i changed them all and the scene lost all references to their prefabs. I was working in my branch called priority and when i switched back to my master branch, its scene was messed up too, so i had to reset all branches to the upstream. This took me 3 times of rewriting the code for this to notice that the errors i was getting were because references was lost because i just was not expecting it.

dustinwloring1988 commented:
I don't see any problems with this, can't wait to have it added.

MANDAL0R3 commented:
I will fix any/all conflicts after StyleCop Saturday is complete.

Mizipzor commented:
Added some floors and ladders, but by right clicking I can only cancel, not prioritize. :/

Mizipzor commented:
I'll merge this but here are some thoughts that popped up during testing.

* Prioritize, as a word, feels like something else. Like, I dont care which character does this, and I certainly dont want to interrupt any current work. But I *do* want it happen before any other queued job.

* This, on the contrary, will force a specific character to do it. Effective use of this will hinge on finding the proper character to force the job with. Is that the wording we should use? Force?

* It also somewhat of a leaky abstraction. Here, the character class manipulated the work queue a little bit too directly. This can be hard to maintain and prone to bugs. I would like to move as much of this as possible from the character code to the job code, specifically the [PrioritizeJob](https://github.com/TeamPorcupine/ProjectPorcupine/pull/825/files#diff-64cc1db7319ecf3cdbc22cea181b644bR271) mehod. Dont get hung up on this, this is hard. :)
